falcon/media/handlers.py

```diff
@@ -39,6 +42,58 @@ def __init__(self, initial=None):
         # Also, this results in self.update(...) being called.
         UserDict.__init__(self, handlers)
 
+    def __setitem__(self, key, item):
+        self.find_by_media_type.cache_clear()
+        super().__setitem__(key, item)
+        self._rehash()
+
+    def __delitem__(self, key):
+        self.find_by_media_type.cache_clear()
+        super().__delitem__(key)
+        self._rehash()
+
+    # NOTE(kgriffs): Make instances hashable so that we can use lru_cache().
+    def __hash__(self):
+        return self._hash
+
+    def _rehash(self):
+        # NOTE(kgriffs): We could have done something simpler, but it would
+        #   have been a leaky abstraction, and it is probably better to avoid
+        #   any potential suprises from instances of Handlers not appearing to
+        #   be logically equivalent.
+        # NOTE(kgriffs): The generator must be wrapped in tuple() in order to
+        #   get a consistent hash value, because otherwise we end up hashing
+        #   the generator object itself.
+        self._hash = hash(tuple((k, v) for k, v in self.data.items()))
```

Comment:
How would this behave in the face of a custom unhashable handler object?

Reply:
Good catch. I did think about unhashable handlers but then it slipped my mind to address it. That being said, a cached closure may do the trick.